File: samples/blog/render.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'fileutils'
require_relative '../../ruby/lib/natsuzora'

BLOG_ROOT = File.expand_path(__dir__)
SHARED_DIR = File.join(BLOG_ROOT, 'shared')
DIST_DIR = File.join(BLOG_ROOT, 'dist')

def render_page(template_path, data_path, output_path)
  source = File.read(template_path)
  data = JSON.parse(File.read(data_path))
  template = Natsuzora::Template.new(source, include_root: SHARED_DIR)
  html = template.render(data)

  FileUtils.mkdir_p(File.dirname(output_path))
  File.write(output_path, html)
rescue StandardError => e
  context = "template=#{template_path}, data=#{data_path}, output=#{output_path}"
  message = "#{e.message}\n  while rendering #{context}"
  raise e.class, message
end

FileUtils.rm_rf(DIST_DIR)

pages = [
  { template: 'index.ntzr', data: 'data.json', output: 'index.html' },
  { template: 'profile.ntzr', data: 'data.json', output: 'profile/index.html' },
  {
    template: 'post.ntzr',
    data: 'post-component-design-best-practices.json',
    output: 'posts/component-design-best-practices/index.html'
  },
  {
    template: 'post.ntzr',
    data: 'post-include-safety-checklist.json',
    output: 'posts/include-safety-checklist/index.html'
  },
  {
    template: 'category.ntzr',
    data: 'category-engineering.json',
    output: 'categories/engineering/index.html'
  }
]

pages.each do |page|
  render_page(
    File.join(BLOG_ROOT, page[:template]),
    File.join(BLOG_ROOT, page[:data]),
    File.join(DIST_DIR, page[:output])
  )
end

puts "Blog sample rendered to #{DIST_DIR}"

File: natsuzora/samples/blog/render.rs
use std::{fs, path::Path};

use natsuzora::Natsuzora;
use serde_json::Value;

fn main() -> natsuzora::Result<()> {
    let root = Path::new(env!("CARGO_MANIFEST_DIR")).join("samples/blog");

    // 1. トップページ
    render_page(
        root.join("index.ntzr"),
        root.join("shared"),
        root.join("data.json"),
        root.join("dist/index.html"),
    )?;

    // 2. プロフィール
    render_page(
        root.join("profile.ntzr"),
        root.join("shared"),
        root.join("data.json"),
        root.join("dist/profile/index.html"),
    )?;

    // 3. 個別記事 (2本)
    render_page(
        root.join("post.ntzr"),
        root.join("shared"),
        root.join("post-component-design-best-practices.json"),
        root.join("dist/posts/component-design-best-practices/index.html"),
    )?;
    render_page(
        root.join("post.ntzr"),
        root.join("shared"),
        root.join("post-include-safety-checklist.json"),
        root.join("dist/posts/include-safety-checklist/index.html"),
    )?;

    // 4. カテゴリ一覧 (例: Engineering)
    render_page(
        root.join("category.ntzr"),
        root.join("shared"),
        root.join("category-engineering.json"),
        root.join("dist/categories/engineering/index.html"),
    )?;

    println!("Blog sample written to {}/dist", root.display());
    Ok(())
}

fn render_page(
    template_path: impl AsRef<Path>,
    include_root: impl AsRef<Path>,
    data_path: impl AsRef<Path>,
    output_path: impl AsRef<Path>,
) -> natsuzora::Result<()> {
    let template = fs::read_to_string(template_path.as_ref())?;
    let data: Value = serde_json::from_slice(&fs::read(data_path.as_ref())?)?;

    let engine = Natsuzora::parse_with_includes(&template, include_root)?;
    let html = engine.render(data)?;

    if let Some(parent) = output_path.as_ref().parent() {
        fs::create_dir_all(parent)?;
    }
    fs::write(output_path, html)?;
    Ok(())
}
